src/lib/context.js

```diff
@@ -57,39 +57,67 @@ export class Context {
           () => langs);
   }
 
-  _resolve(langs, id, args, formatter) {
+  _reportMissing(missingIdArgs, formatter, resolved) {
+    const missingIds = new Set();
+
+    missingIdArgs.forEach((idArgs, i) => {
+      if (i in resolved) {
+        return;
+      }
+      const id = Array.isArray(idArgs) ? idArgs[0] : idArgs;
+      missingIds.add(id);
+      resolved[i] = formatter === this._formatValue ?
+        id : {value: id, args: null};
+    });
+
+    this._env.emit('notfounderror', new L10nError(
+      '"' + [...missingIds].join(', ') + '"' +
+      ' not found in any language', missingIds), this);
+
+    return resolved;
+  }
+
+  _resolve(langs, idArgsList, formatter, resolved = []) {
     const lang = langs[0];
 
     if (!lang) {
-      this._env.emit('notfounderror', new L10nError(
-        '"' + id + '"' + ' not found in any language', id), this);
-      if (formatter === this._formatEntity) {
-        return { value: id, attrs: null };
-      } else {
-        return id;
-      }
+      return this._reportMissing(idArgsList, formatter, resolved);
     }
 
-    const entity = this._getEntity(lang, id);
+    var hasUnresolved = false;
+
+    idArgsList.forEach((idArgs, i) => {
+      if (i in resolved) {
+        return;
+      }
+      const id = Array.isArray(idArgs) ? idArgs[0] : idArgs;
+      const args = Array.isArray(idArgs) ? idArgs[1] : undefined;
+      const entity = this._getEntity(lang, id);
 
-    if (entity) {
-      return Promise.resolve(
-        formatter.call(this, lang, args, entity, id));
-    } else {
-      this._env.emit('notfounderror', new L10nError(
-        '"' + id + '"' + ' not found in ' + lang.code, id, lang), this);
+      if (entity) {
+        resolved[i] = formatter.call(this, lang, args, entity, id);
+      } else {
+        this._env.emit('notfounderror',
+          new L10nError('"' + id + '"' + ' not found in ' + lang.code,
+            id, lang), this);
+        hasUnresolved = true;
+      }
+    });
+
+    if (!hasUnresolved) {
+      return Promise.all(resolved);
```

Comment:
Is `resolved` an array of promises?  `formatter` should return synchronous values, in which case you should be able to `return Promise.resolve(resolved)` here.  Note that this should much faster:  it's only one promise vs. one promise per element of `resolved`.